openedx/core/djangoapps/xblock/olx_bundle/__init__.py

```diff
@@ -0,0 +1,346 @@
+        have no parent. (Excludes blocks from linked bundles.)
+        """
+        top_level_definitions, _ua = self._scan_bundle_olx()
+        return top_level_definitions
+
+    def _scan_bundle_olx(self):
+        """
+        Helper method to scan bundles that follow certain conventions and return
+        both a list of "top level" blocks and a map of usage IDs to definitions.
+
+        ** In the future this method can/should be optimized away by performing
+           this computation on commit and writing the result into an olx.map
+           file within the bundle itself **
+
+        ** This method follows the semantics of course and library bundles, but
+           other types of bundles may not follow the same patterns. **
+
+        A course or library bundle consists of OLX files that MUST be named
+            {block_type}/{definition_id}/definition.xml
+        For example:
+            html/intro/definition.xml
+            unit/unit1/definition.xml
+        And it may also have linked bundles which contain additional OLX like:
+            linked_bundle/html/intro/definition.xml
+            linked_bundle/second_link/html/intro/definition.xml
+
+        Any OLX file that represents an XBlock type that can have children may
+        contain <xblock-include> nodes, like this:
+            <unit>
+                <xblock-include definition="html/intro" />
+                <xblock-include definition="html/intro" usage="alernate" />
+                <xblock-include definition="linked_bundle/html/intro" usage="linked_intro" />
+            </unit>
+
+        If the included child is from a different bundle, a usage key *must* be
+        specified (this greatly simplifies the implementation and reduces
+        potential namespace collisions).
+
+        If a block is included via an <xblock-include> directive with a
+        usage="..." alias, and the included block has children, the usage IDs of
+        the child blocks (and their children, and so on) will be prefixed with
+        the usage alias. (Otherwise, if that block was used in multiple places
+        in the same bundle, it would have a unique usage ID in each case, but
+        its children would not.)
+
+        Any block which exists in the main bundle (not linked in to the bundle)
+        and which is not a child of other blocks (no <xblock-include /> nodes
+        reference it) is called a "top-level block". Its usage ID always matches
+        its definition ID.
+
+        Any other blocks ("child blocks") will have at least one parent/ancestor
+        in the main bundle, and possibly more. Their usage ID may or may not
+        match their definition ID depending on whether or not a usage="blah"
+        parameter was used in their parent's <xblock-include> element.
+
+        So, given a usage ID like 'intro', the following algorithm is used to
+        determine which definition it belongs to:
```

Comment:
Doesn't the "usage ID" contain the definition path already? Or maybe I am not understanding what this is?

Reply:
In this PR a usage ID is an opaque key like `lb:library-id:html:intro` which doesn't necessarily encode the definition path.

If no `<xblock-include>` directives with `usage="..."` exist in the bundle, then yes, the usage ID always directly relates to the definition, so `lb:library-id:html:intro` would refer to the definition at `html/intro/definition.xml`.

However, if someone wants to create a second usage of that definition, they would write `<xblock-include definition="html/intro" usage="other">`. That results in the opaque usage key `lb:library-id:html:other` which you can't directly map to a definition without inspecting the bundle.

The whole reason for this is to allow authors to swap the definition without changing the usage ID (and thus the student state). But now that I think about that, maybe it's not needed, and we could just always encode the definition ID _and_ the usage ID in the key? That would certainly make this part a lot more efficient and avoid all these bundle scans. It would take away the option to swap the underlying definition used at any particular usage point in the course/library though.

Reply:
Hmm. I am trying to understand how this alternative usage functionality will be used by xblocks executing in LX. What will be the keys of standalone XBlocks and what will be the keys of XBlocks that are part of a Pathway and how will they map to this usage functionality?

Reply:
This won't really apply to LX.

In LX, each standalone XBlock will be created in a library (bundle), as a single definition with a single usage. There may be one or multiple blocks per bundle (prob. just one block per bundle at first since Blockstore doesn't yet support the partial draft/publish workflow we'd need to properly store multiple blocks per bundle).

Then, there are two options for how we implement pathways:
* Purely in LX: each pathway is a list of usage IDs stored in LX. This is more or less how it already has been implemented. This provides that each XBlock has a consistent state, grade, and completion status regardless of which pathway the user has viewed it in. But the downside of this approach is that pathway authors can't pick a specific version of each XBlock, because there's no "Link" to freeze the block to a specific version. So updates to the original block apply immediately to every pathway.
    - In this case there would only ever be one usage per definition in LX, though the content library feature would support multiple usages for definition for other edX use cases.
* Pathway bundles: each pathway is a bundle, and every block in the pathway has its bundle "linked" into the pathway bundle using blockstore links. There is some file which lists all the blocks that are in the pathway, creating a new pathway-specific usage of each of those block definitions. This provides each XBlock a new state, grade, and completion status within this pathway, and also freezes the version of the block to be the version that the original author chose.
    - But these are inter-bundle usages not intra-bundle usages. The original library bundles still have only one usage per definition within their bundle.

Reply:
> The whole reason for this is to allow authors to swap the definition without changing the usage ID (and thus the student state). But now that I think about that, maybe it's not needed, and we could just always encode the definition ID and the usage ID in the key?

@ormsbee Do we have a compelling reason to allow authors to swap the underlying definition used by a block in a library, keeping the same usage key? I thought that was important but now it seems like an unnecessary complicating factor.

Reply:
> The whole reason for this is to allow authors to swap the definition without changing the usage ID (and thus the student state). But now that I think about that, maybe it's not needed, and we could just always encode the definition ID and the usage ID in the key? That would certainly make this part a lot more efficient and avoid all these bundle scans. It would take away the option to swap the underlying definition used at any particular usage point in the course/library though.

> Do we have a compelling reason to allow authors to swap the underlying definition used by a block in a library, keeping the same usage key?

Ok, there is a substantial earlier discussion about this here: https://github.com/open-craft/edx-platform/pull/146#discussion_r228612958 I guess I am re-treading an earlier discussion.

Some key points are:

On *not* encoding the definition location in the usage key:

>  I could see things moving around pretty easily -- files getting renamed to more appropriate things, items being moved between units, etc. And I'd be concerned if we were storing any user state against that kind of definition reference.

The thing here is, because our bundle file format is now pretty strict about file locations, "files getting renamed to more appropriate things" really just means "changing the ID of a definition". So yeah, we _can_ support that case (and I currently have), but things could be a lot faster and simpler if we don't have to.

> I think that in the v1, this problem is manageable without any special facilities on Blockstore's side. As long as we constrain Bundles in practice to be used on the scale of single units or sequences, then reading all the OLX and processing it together once you've got it narrowed down to the right Bundle should be fine.

A couple things that have changed since this discussion though is:  
(1) I haven't built any `bundle.json` or similar top-level metadata file into the content library v2 bundle format (for two reasons: because it's a potential source of inconsistent data if not kept perfectly in sync with the bundle contents, and because we want authors to be able to publish changes to individual blocks, which we can do if they're in their own folder but cannot do if it requires changes to some lines in a bundle.json file but not other lines)
(2) I built content library bundles to support many blocks per bundle if that's the appropriate granularity, unlike the previous one block per bundle library idea. This is more consistent with courses. So now for example, if creating a single asset for a use case like LabXchange, it can get its own library/bundle, but if creating a problem bank, all the problems can be in a single library/bundle.


Here are my current thoughts:
* The primary purpose of a content library is to contain XBlock _definitions_ that authors can use in courses and other libraries.
    - Although it's useful to have a usage key for each block in the library in order to allow learners to "learn" from the library directly (LX use case), there is no need for libraries to support "Swappable" usage keys where usage key "problem1" in the library first points to definition "problemA" but then is later changed to completely different "problemB".
        - This implies that for libraries, the usage key can directly encode the definition ID
        - This feature could be useful for courses though, e.g. if a course author wants to move the underlying definition out of the course and into a library while keeping the usage ID unchanged?
    - A single definition should be allowed to appear (as a child) in multiple places in the library although when used by students directly or via another course/library, each distinct use of the same definition and its children must have unique usage keys.
        - However, the case where one definition appears in multiple places in the same library is expected to be extremely rare. One library definition linked into multiple places in the same course (different bundle) will be fairly common though (e.g. randomized problem assignments throughout the course happen to randomly select the same question).
* Most libraries will largely be a collection of single-block definitions, although sometimes definitions will have other definitions as children. If so, it's important to know whether or not those children should be discoverable and reusable independently of their parent.
    - What I've done for now, to avoid a bundle.json or other stateful file, is to say that any block which gets used as a child anywhere in the library bundle is not meant for discoverability on its own, and must be discovered via its parent. However, once you discover it you may certainly reuse it in a course on its own.
        - This implies that a block cannot be discoverable on its own ("top level") and a child of other blocks in the library. I think that's fine? I can imagine an author creating a `problem` block in a library then deciding to _move_ it into a `vertical` alongside another `problem` but I don't really foresee an author creating a problem block and then wanting it staying as a top level block while also being linked into some other component in the library?
        - What I like about this is that it cannot get into an inconsistent state and it's impossible to have "orphaned" definitions that exist in the library but are not discoverable
        - What I don't like is that this requires scanning every OLX file in the bundle to read all the `<xblock-include>` directives in order to determine what the top-level blocks are. While this can be cached, if the cache is lost or expires, the request to "list all top level blocks" can then be slow.
        - One optimization is in place: there is no need to read .olx files of any block type known to not support child blocks, so in practice any library that doesn't contain `vertical`, `unit`, or `problem-builder` blocks is still very fast.
        - I have implemented this along with a cache but the cache is in memcache so it can expire and the first load is still potentially slow enough to time out if there is no limit on the number of blocks with children allowed in the library.

So what I'd like to do is:
* Change the library usage keys to directly encode the definition, greatly simplifying the usage->definition lookup
* Find a better way to label specific blocks as "top level" / discoverable apart from their parent. Or ideally preserve the existing case (blocks with no parent are "top level", blocks cannot be "top level" and appear as children in the library) but avoid the case where a library with hundreds of `problem-builder` blocks that each have 3-5 child xblocks would take a very long time to scan.

Reply:
> @ormsbee Do we have a compelling reason to allow authors to swap the underlying definition used by a block in a library, keeping the same usage key? I thought that was important but now it seems like an unnecessary complicating factor.

Maybe because it's crossing what we mean by "definitions"? In the Modulestore world, the usage key is what stays stable for the purposes of storing student state, but if you change a content scoped field, you end up with an entirely new definition (since entries in the definitions collection are immutable).

Reply:
Yeah, here by "definition" I mean "the block as authored and maintained by the author, with Scope.content and Scope.settings values", that may then be used in multiple different _usages_ in various courses and libraries. (Maaayyybeee with different Scope.settings overrides per-usage). Or another way to put it is the "definition" is the OLX file, and the usage is what learners see in a particular learning context.

Reply:
> So what I'd like to do is:
> 
> * Change the library usage keys to directly encode the definition, greatly simplifying the usage->definition lookup
> * Find a better way to label specific blocks as "top level" / discoverable apart from their parent. Or ideally preserve the existing case (blocks with no parent are "top level", blocks cannot be "top level" and appear as children in the library) but avoid the case where a library with hundreds of problem-builder blocks that each have 3-5 child xblocks would take a very long time to scan.

Okay, I have a much longer reply coming, but I agree with both of these points. Some side thoughts:

* There's a lot of complexity in the system that comes from the usage/definition divide, and I sometimes wonder if we can just unify that with something simpler without negating the whole point of linking/sharing.
* Declaring that something is intended for advertised reuse could be putting a flag at the individual OLX file level and having something more efficient index it for later querying. The Bundle format should support easy lookup and the basic read/write operations, but I don't think we should have to accommodate search use cases.

Reply:
Hmm, should I just dump the `block.display_name`, `block.student_view_user_state`, `block.index_dictionary`, and `block.children` into ElasticSearch ? That would be pretty easy and make everything fast. I've already done something similar on the LX backend which uses ElasticSearch heavily.

Reply:
Looking forward to your longer reply :)

Reply:
> Hmm, should I just dump the block.display_name, block.student_view_user_state, block.index_dictionary, and block.children into ElasticSearch ? That would be pretty easy and make everything fast. I've already done something similar on the LX backend which uses ElasticSearch heavily.

I think it gets messier if you're talking Drafts, but if you're just talking about immutable BundleVersions, then yeah, stick all that stuff in and make an app that queries it efficiently and surfaces whatever we care about. We get all the facet-searching power we want while keeping the Bundle format stupid and optimized for "find the thing corresponding to this key so I can edit it".

On that note, since we've agreed before that the Bundle format is for machine-reading and not optimized for humans, I think it's fine to always require an explicit usage param for includes.

Reply:
> There's a lot of complexity in the system that comes from the usage/definition divide, and I sometimes wonder if we can just unify that with something simpler without negating the whole point of linking/sharing.

After thinking this over, what I'd like to do in this code going forward, is to use a really simple definition of "definition":

```
definition = a specific .olx file in a particular bundleversion
```

Anything else more abstract than "a specific .olx file" (e.g. a block in a library, block in a course, etc.) is a "usage" that is backed by some definition.

I think I now know how I need to refactor this code (fielddata implementation in particular) to make it much simpler and to support links; working on that now.

Reply:
Ok, I realized why we probably shouldn't put the definition IDs (olx file path) in the usage keys.

If the usage key encodes the definition (the path to the underlying OLX file), then: the usage key must also encode the link information (is the definition/olxfile in the same bundle, or in a linked bundle)

Consequently:
* you can't move the definition (olx file) out of the bundle and into a linked bundle, without changing the usage key (minor concern?)
* if the definition is in a linked bundle and gets renamed, or e.g. a linked bundle of problems is split into two separate bundles, one cannot update the usage to point to the new definition without changing the usage key (seems like a more serious concern)